File: Recursion/Lecture 10/CodeStudio/Subsets II.cpp
#include <bits/stdc++.h>
void helper(int i, vector<int> &arr, vector<int> &subset, vector<vector<int>> &ans)
{
    // when we have traversed the entire arr[], then we should put arr into ans[]
    if (i >= arr.size())
    {
        ans.push_back(subset);
        return;
    }

    // pick the ith element
    subset.push_back(arr[i]);

    // picked ith element and recursion will find the remaining subsets with ith element
    helper(i + 1, arr, subset, ans);

    // firstly, we need to backtrack and undo the changes that we have made
    subset.pop_back();

    // ensure we don't pick any other occurrences ofi-th element in order to generate Unique Subsets 
    while (i + 1 < arr.size() && arr[i] == arr[i + 1]) i++;

    // currentIndex will now point to the Last Occurrence of ith Element
    // we ignored ith element and now recursion will do find the remaining subsets without the ith element
    // skip the ith element
    helper(i + 1, arr, subset, ans);
}

vector<vector<int>> uniqueSubsets(int n, vector<int> &arr)
{
    // Write your code here.
    vector<vector<int>> ans;
    vector<int> subset;

    // to ensure our logic work,we need a sorted array
    sort(arr.begin(), arr.end());
    helper(0, arr, subset, ans);

    // this is specific to the given problem, they wanted each subset in sorted order
    sort(ans.begin(), ans.end());
    return ans;
}

/*
    time complexity : O(2^n)
    space complexity : O(n)
*/

File: Recursion/Lecture 10/CodeStudio/Subsets II.py
from typing import *

def uniqueSubsets(n :int,arr :List[int]) -> List[List[int]]:

    def helper(i, subset, arr, ans):
        
        #Base Case, When we have reached the end of nums[] , we have generated a Valid Unique Subset for sure
        
        if i >= len(arr):
            ans.append(subset.copy())
            return
        
        # pick the ith element
        subset.append(arr[i])
        
        #  picked ith element and recursion will find the remaining subsets with ith element | Left Subtree
        
        helper(i+1, subset, arr, ans)
        
        # firstly, we need to backtrack and undo the changes that we have made
        
        subset.pop()
        
        # ensure we don't pick any other occurrences ofi-th element in order to generate Unique Subsets
        
        while i+1 < len(arr) and arr[i] == arr[i+1]:#Preventon step
            i += 1
            
        # currentIndex will now point to the Last Occurrence of ith Element
        # we ignored ith element and now recursion will do find the remaining subsets without the ith element
        # skip the ith element
        
        helper(i+1, subset, arr, ans)
    
    # For our logic to work, we need to sort the given nums[] vector
    arr.sort()
    
    ans = []
    
    subset= []
    
    helper(0, subset, arr, ans)
    
    # this is specific to the given problem, they wanted each subset in sorted order
    ans.sort()
    return ans
    
"""
Time Complexity  : O(2^N)
Space Complexity : O(N)
"""
